File: .skills/openclaw-skills/skills/0xnyk/xint/lib/health.ts
/**
 * lib/health.ts — runtime health and auth diagnostics.
 */

import { readFileSync } from "fs";
import { join } from "path";
import { BASE } from "./api";
import { checkBudget, getTodayCosts } from "./costs";
import { getReliabilityReport } from "./reliability";
import { loadTokens } from "./oauth";

type StatusLevel = "ok" | "warn" | "fail";

interface ServiceCheck {
  configured: boolean;
  valid: boolean;
  status: StatusLevel;
  message: string;
  latency_ms: number | null;
  status_code: number | null;
}

interface OAuthCheck {
  configured: boolean;
  valid: boolean;
  status: StatusLevel;
  username: string | null;
  expires_in_minutes: number | null;
  missing_scopes: string[];
  message: string;
  status_code: number | null;
}

interface AuthDoctorReport {
  checked_at: string;
  bearer: ServiceCheck;
  xai: ServiceCheck;
  oauth: OAuthCheck;
  overall_status: StatusLevel;
}

interface HealthReport {
  checked_at: string;
  overall_status: StatusLevel;
  auth: AuthDoctorReport;
  budget: {
    allowed: boolean;
    warning: boolean;
    spent_usd: number;
    limit_usd: number;
    remaining_usd: number;
  };
  today: {
    calls: number;
    tweets_read: number;
    total_cost_usd: number;
  };
  reliability: ReturnType<typeof getReliabilityReport>;
}

const ENV_FILE = join(import.meta.dir, "..", ".env");
const REQUIRED_OAUTH_SCOPES = [
  "bookmark.read",
  "bookmark.write",
  "like.read",
  "like.write",
  "follows.read",
  "follows.write",
  "list.read",
  "list.write",
  "block.read",
  "block.write",
  "mute.read",
  "mute.write",
  "tweet.read",
  "tweet.write",
  "users.read",
  "offline.access",
];

function combineStatus(current: StatusLevel, next: StatusLevel): StatusLevel {
  const rank: Record<StatusLevel, number> = { ok: 1, warn: 2, fail: 3 };
  return rank[next] > rank[current] ? next : current;
}

function parseEnvFile(): Record<string, string> {
  try {
    const raw = readFileSync(ENV_FILE, "utf-8");
    const out: Record<string, string> = {};
    for (const line of raw.split(/\r?\n/)) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) continue;
      const eq = trimmed.indexOf("=");
      if (eq <= 0) continue;
      const key = trimmed.slice(0, eq).trim();
      const valueRaw = trimmed.slice(eq + 1).trim();
      const value = valueRaw.replace(/^['"]|['"]$/g, "");
      out[key] = value;
    }
    return out;
  } catch {
    return {};
  }
}

function getConfigValue(name: string): string | undefined {
  if (process.env[name]) return process.env[name];
  const env = parseEnvFile();
  return env[name] || undefined;
}

async function checkBearerToken(token: string | undefined): Promise<ServiceCheck> {
  if (!token) {
    return {
      configured: false,
      valid: false,
      status: "fail",
      message: "X_BEARER_TOKEN missing",
      latency_ms: null,
      status_code: null,
    };
  }

  const started = Date.now();
  const url = `${BASE}/users/by/username/x?user.fields=id`;

  try {
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}` },
    });

    return {
      configured: true,
      valid: res.ok,
      status: res.ok ? "ok" : "fail",
      message: res.ok ? "Bearer token valid" : `Bearer token check failed (${res.status})`,
      latency_ms: Date.now() - started,
      status_code: res.status,
    };
  } catch (error) {
    return {
      configured: true,
      valid: false,
      status: "warn",
      message: `Bearer token check error: ${error instanceof Error ? error.message : String(error)}`,
      latency_ms: Date.now() - started,
      status_code: null,
    };
  }
}

async function checkXaiToken(token: string | undefined): Promise<ServiceCheck> {
  if (!token) {
    return {
      configured: false,
      valid: false,
      status: "warn",
      message: "XAI_API_KEY missing (AI features disabled)",
      latency_ms: null,
      status_code: null,
    };
  }

  const started = Date.now();
  const url = "https://api.x.ai/v1/models";

  try {
    const res = await fetch(url, {
      headers: { Authorization: `Bearer ${token}` },
    });

    return {
      configured: true,
      valid: res.ok,
      status: res.ok ? "ok" : "fail",
      message: res.ok ? "xAI key valid" : `xAI key check failed (${res.status})`,
      latency_ms: Date.now() - started,
      status_code: res.status,
    };
  } catch (error) {
    return {
      configured: true,
      valid: false,
      status: "warn",
      message: `xAI key check error: ${error instanceof Error ? error.message : String(error)}`,
      latency_ms: Date.now() - started,
      status_code: null,
    };
  }
}

async function checkOAuthTokens(): Promise<OAuthCheck> {
  const tokens = loadTokens();
  if (!tokens) {
    return {
      configured: false,
      valid: false,
      status: "warn",
      username: null,
      expires_in_minutes: null,
      missing_scopes: REQUIRED_OAUTH_SCOPES,
      message: "OAuth tokens missing (engagement/moderation commands unavailable)",
      status_code: null,
    };
  }

  const expiresInMinutes = Math.floor((tokens.expires_at - Date.now()) / 60_000);
  const tokenScopes = new Set(tokens.scope.split(/\s+/).filter(Boolean));
  const missingScopes = REQUIRED_OAUTH_SCOPES.filter((scope) => !tokenScopes.has(scope));

  try {
    const res = await fetch(`${BASE}/users/me`, {
      headers: { Authorization: `Bearer ${tokens.access_token}` },
    });

    const valid = res.ok;
    let status: StatusLevel = "ok";
    let message = valid ? "OAuth access token valid" : `OAuth token check failed (${res.status})`;

    if (expiresInMinutes <= 0) {
      status = "warn";
      message = "OAuth access token expired (refresh required)";
    }
    if (missingScopes.length > 0) {
      status = combineStatus(status, "warn");
      message = `${message}; missing scopes: ${missingScopes.join(", ")}`;
    }
    if (!valid) {
      status = "fail";
    }

    return {
      configured: true,
      valid,
      status,
      username: tokens.username,
      expires_in_minutes: expiresInMinutes,
      missing_scopes: missingScopes,
      message,
      status_code: res.status,
    };
  } catch (error) {
    return {
      configured: true,
      valid: false,
      status: "warn",
      username: tokens.username,
      expires_in_minutes: expiresInMinutes,
      missing_scopes: missingScopes,
      message: `OAuth token check error: ${error instanceof Error ? error.message : String(error)}`,
      status_code: null,
    };
  }
}

async function collectAuthDoctorReport(): Promise<AuthDoctorReport> {
  const bearer = await checkBearerToken(getConfigValue("X_BEARER_TOKEN"));
  const xai = await checkXaiToken(getConfigValue("XAI_API_KEY"));
  const oauth = await checkOAuthTokens();

  let overall: StatusLevel = "ok";
  overall = combineStatus(overall, bearer.status);
  overall = combineStatus(overall, xai.status);
  overall = combineStatus(overall, oauth.status);

  return {
    checked_at: new Date().toISOString(),
    bearer,
    xai,
    oauth,
    overall_status: overall,
  };
}

function parseWindowDays(args: string[]): number {
  const idx = args.indexOf("--days");
  if (idx < 0 || idx + 1 >= args.length) return 7;
  const raw = parseInt(args[idx + 1], 10);
  if (Number.isNaN(raw)) return 7;
  return Math.max(1, Math.min(30, raw));
}

export async function cmdAuthDoctor(args: string[]): Promise<void> {
  const asJson = args.includes("--json");
  const report = await collectAuthDoctorReport();

  if (asJson) {
    console.log(JSON.stringify(report, null, 2));
    return;
  }

  const icon = (status: StatusLevel) => status === "ok" ? "✅" : status === "warn" ? "⚠️" : "❌";
  console.log(`${icon(report.overall_status)} Auth Doctor (${report.checked_at})`);
  console.log(`- Bearer: ${icon(report.bearer.status)} ${report.bearer.message}`);
  console.log(`- xAI: ${icon(report.xai.status)} ${report.xai.message}`);
  const oauthUser = report.oauth.username ? `@${report.oauth.username}` : "not configured";
  console.log(`- OAuth: ${icon(report.oauth.status)} ${report.oauth.message} (${oauthUser})`);
}

export async function cmdHealth(args: string[]): Promise<void> {
  const asJson = args.includes("--json");
  const windowDays = parseWindowDays(args);
  const auth = await collectAuthDoctorReport();
  const budget = checkBudget();
  const today = getTodayCosts();
  const reliability = getReliabilityReport(windowDays);

  let overall: StatusLevel = auth.overall_status;
  if (!budget.allowed) {
    overall = combineStatus(overall, "warn");
  }
  if (reliability.total_calls > 0 && reliability.success_rate < 0.9) {
    overall = combineStatus(overall, "warn");
  }

  const report: HealthReport = {
    checked_at: new Date().toISOString(),
    overall_status: overall,
    auth,
    budget: {
      allowed: budget.allowed,
      warning: budget.warning,
      spent_usd: budget.spent,
      limit_usd: budget.limit,
      remaining_usd: budget.remaining,
    },
    today: {
      calls: today.calls,
      tweets_read: today.tweets_read,
      total_cost_usd: today.total_cost,
    },
    reliability,
  };

  if (asJson) {
    console.log(JSON.stringify(report, null, 2));
    return;
  }

  const icon = (status: StatusLevel) => status === "ok" ? "✅" : status === "warn" ? "⚠️" : "❌";
  console.log(`${icon(report.overall_status)} xint Health (${report.checked_at})`);
  console.log(`- Auth status: ${icon(report.auth.overall_status)} ${report.auth.overall_status}`);
  console.log(`- Budget: ${report.budget.allowed ? "within limit" : "exceeded"} ($${report.budget.spent_usd.toFixed(2)} / $${report.budget.limit_usd.toFixed(2)})`);
  console.log(`- Reliability (${windowDays}d): ${(report.reliability.success_rate * 100).toFixed(1)}% success across ${report.reliability.total_calls} calls`);

  const topCommands = Object.entries(report.reliability.by_command)
    .sort((a, b) => b[1].calls - a[1].calls)
    .slice(0, 5);

  if (topCommands.length > 0) {
    console.log("- Top commands:");
    for (const [command, stats] of topCommands) {
      console.log(
        `  ${command}: ${(stats.success_rate * 100).toFixed(1)}% ok, p95 ${stats.p95_latency_ms.toFixed(0)}ms, fallback ${(stats.fallback_rate * 100).toFixed(1)}%`,
      );
    }
  }
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/health.rs
use anyhow::Result;
use serde::Serialize;

use crate::auth::oauth;
use crate::cli::HealthArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::format;
use crate::reliability;

#[derive(Debug, Clone, Copy, Serialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
enum StatusLevel {
    Ok,
    Warn,
    Fail,
}

#[derive(Debug, Clone, Serialize)]
struct ServiceCheck {
    configured: bool,
    valid: bool,
    status: StatusLevel,
    message: String,
    latency_ms: Option<u128>,
    status_code: Option<u16>,
}

#[derive(Debug, Clone, Serialize)]
struct OAuthCheck {
    configured: bool,
    valid: bool,
    status: StatusLevel,
    username: Option<String>,
    expires_in_minutes: Option<i64>,
    missing_scopes: Vec<String>,
    message: String,
    status_code: Option<u16>,
}

#[derive(Debug, Clone, Serialize)]
pub struct AuthDoctorReport {
    checked_at: String,
    bearer: ServiceCheck,
    xai: ServiceCheck,
    oauth: OAuthCheck,
    overall_status: StatusLevel,
}

#[derive(Debug, Clone, Serialize)]
struct BudgetView {
    allowed: bool,
    warning: bool,
    spent_usd: f64,
    limit_usd: f64,
    remaining_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
struct TodayView {
    calls: u64,
    tweets_read: u64,
    total_cost_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
struct HealthReport {
    checked_at: String,
    overall_status: StatusLevel,
    auth: AuthDoctorReport,
    budget: BudgetView,
    today: TodayView,
    reliability: reliability::ReliabilityReport,
}

const REQUIRED_OAUTH_SCOPES: &[&str] = &[
    "bookmark.read",
    "bookmark.write",
    "like.read",
    "like.write",
    "follows.read",
    "follows.write",
    "list.read",
    "list.write",
    "block.read",
    "block.write",
    "mute.read",
    "mute.write",
    "tweet.read",
    "tweet.write",
    "users.read",
    "offline.access",
];

fn combine_status(current: StatusLevel, next: StatusLevel) -> StatusLevel {
    let rank = |s: StatusLevel| match s {
        StatusLevel::Ok => 1,
        StatusLevel::Warn => 2,
        StatusLevel::Fail => 3,
    };
    if rank(next) > rank(current) {
        next
    } else {
        current
    }
}

fn icon(status: StatusLevel) -> &'static str {
    match status {
        StatusLevel::Ok => "✅",
        StatusLevel::Warn => "⚠️",
        StatusLevel::Fail => "❌",
    }
}

async fn check_bearer(config: &Config, client: &XClient) -> ServiceCheck {
    let token = match config.bearer_token.as_deref() {
        Some(token) => token,
        None => {
            return ServiceCheck {
                configured: false,
                valid: false,
                status: StatusLevel::Fail,
                message: "X_BEARER_TOKEN missing".to_string(),
                latency_ms: None,
                status_code: None,
            }
        }
    };

    let started = std::time::Instant::now();
    match client
        .bearer_get("users/by/username/x?user.fields=id", token)
        .await
    {
        Ok(_) => ServiceCheck {
            configured: true,
            valid: true,
            status: StatusLevel::Ok,
            message: "Bearer token valid".to_string(),
            latency_ms: Some(started.elapsed().as_millis()),
            status_code: Some(200),
        },
        Err(err) => ServiceCheck {
            configured: true,
            valid: false,
            status: StatusLevel::Fail,
            message: format!("Bearer token check failed: {err}"),
            latency_ms: Some(started.elapsed().as_millis()),
            status_code: None,
        },
    }
}

async fn check_xai(config: &Config) -> ServiceCheck {
    let key = match config.xai_api_key.as_deref() {
        Some(key) => key,
        None => {
            return ServiceCheck {
                configured: false,
                valid: false,
                status: StatusLevel::Warn,
                message: "XAI_API_KEY missing (AI features disabled)".to_string(),
                latency_ms: None,
                status_code: None,
            }
        }
    };

    let started = std::time::Instant::now();
    let http = reqwest::Client::new();
    match http
        .get("https://api.x.ai/v1/models")
        .bearer_auth(key)
        .send()
        .await
    {
        Ok(res) => {
            let status = res.status().as_u16();
            if res.status().is_success() {
                ServiceCheck {
                    configured: true,
                    valid: true,
                    status: StatusLevel::Ok,
                    message: "xAI key valid".to_string(),
                    latency_ms: Some(started.elapsed().as_millis()),
                    status_code: Some(status),
                }
            } else {
                ServiceCheck {
                    configured: true,
                    valid: false,
                    status: StatusLevel::Fail,
                    message: format!("xAI key check failed ({status})"),
                    latency_ms: Some(started.elapsed().as_millis()),
                    status_code: Some(status),
                }
            }
        }
        Err(err) => ServiceCheck {
            configured: true,
            valid: false,
            status: StatusLevel::Warn,
            message: format!("xAI key check error: {err}"),
            latency_ms: Some(started.elapsed().as_millis()),
            status_code: None,
        },
    }
}

async fn check_oauth(config: &Config, client: &XClient) -> OAuthCheck {
    let tokens = match oauth::load_tokens(&config.tokens_path()) {
        Some(tokens) => tokens,
        None => {
            return OAuthCheck {
                configured: false,
                valid: false,
                status: StatusLevel::Warn,
                username: None,
                expires_in_minutes: None,
                missing_scopes: REQUIRED_OAUTH_SCOPES
                    .iter()
                    .map(|scope| scope.to_string())
                    .collect(),
                message: "OAuth tokens missing (engagement/moderation commands unavailable)"
                    .to_string(),
                status_code: None,
            }
        }
    };

    let now_ms = chrono::Utc::now().timestamp_millis();
    let expires_in_minutes = (tokens.expires_at - now_ms) / 60_000;

    let scope_set: std::collections::HashSet<_> = tokens.scope.split_whitespace().collect();
    let missing_scopes: Vec<String> = REQUIRED_OAUTH_SCOPES
        .iter()
        .filter(|scope| !scope_set.contains(**scope))
        .map(|scope| scope.to_string())
        .collect();

    let (mut status, valid, mut message, status_code) =
        match client.oauth_get("users/me", &tokens.access_token).await {
            Ok(_) => (
                StatusLevel::Ok,
                true,
                "OAuth access token valid".to_string(),
                Some(200),
            ),
            Err(err) => (
                StatusLevel::Fail,
                false,
                format!("OAuth token check failed: {err}"),
                None,
            ),
        };

    if expires_in_minutes <= 0 {
        status = combine_status(status, StatusLevel::Warn);
        message = "OAuth access token expired (refresh required)".to_string();
    }

    if !missing_scopes.is_empty() {
        status = combine_status(status, StatusLevel::Warn);
        message = format!("{message}; missing scopes: {}", missing_scopes.join(", "));
    }

    OAuthCheck {
        configured: true,
        valid,
        status,
        username: Some(tokens.username),
        expires_in_minutes: Some(expires_in_minutes),
        missing_scopes,
        message,
        status_code,
    }
}

fn status_label(status: StatusLevel) -> &'static str {
    match status {
        StatusLevel::Ok => "ok",
        StatusLevel::Warn => "warn",
        StatusLevel::Fail => "fail",
    }
}

pub async fn auth_doctor_report(config: &Config, client: &XClient) -> AuthDoctorReport {
    let bearer = check_bearer(config, client).await;
    let xai = check_xai(config).await;
    let oauth = check_oauth(config, client).await;

    let mut overall = StatusLevel::Ok;
    overall = combine_status(overall, bearer.status);
    overall = combine_status(overall, xai.status);
    overall = combine_status(overall, oauth.status);

    AuthDoctorReport {
        checked_at: chrono::Utc::now().to_rfc3339(),
        bearer,
        xai,
        oauth,
        overall_status: overall,
    }
}

pub async fn run_auth_doctor(
    config: &Config,
    client: &XClient,
    as_json: bool,
) -> Result<AuthDoctorReport> {
    let report = auth_doctor_report(config, client).await;

    if as_json {
        format::print_json_pretty_filtered(&report)?;
    } else {
        println!(
            "{} Auth Doctor ({})",
            icon(report.overall_status),
            report.checked_at
        );
        println!(
            "- Bearer: {} {}",
            icon(report.bearer.status),
            report.bearer.message
        );
        println!("- xAI: {} {}", icon(report.xai.status), report.xai.message);
        let oauth_user = report
            .oauth
            .username
            .as_ref()
            .map(|name| format!("@{name}"))
            .unwrap_or_else(|| "not configured".to_string());
        println!(
            "- OAuth: {} {} ({oauth_user})",
            icon(report.oauth.status),
            report.oauth.message
        );
    }

    Ok(report)
}

pub async fn run(args: &HealthArgs, config: &Config, client: &XClient) -> Result<()> {
    let auth = auth_doctor_report(config, client).await;
    let budget = costs::check_budget(&config.costs_path());
    let today = costs::today_costs(&config.costs_path());
    let reliability = reliability::get_reliability_report(&config.reliability_path(), args.days);

    let mut overall = auth.overall_status;
    if !budget.allowed {
        overall = combine_status(overall, StatusLevel::Warn);
    }
    if reliability.total_calls > 0 && reliability.success_rate < 0.9 {
        overall = combine_status(overall, StatusLevel::Warn);
    }

    let report = HealthReport {
        checked_at: chrono::Utc::now().to_rfc3339(),
        overall_status: overall,
        auth,
        budget: BudgetView {
            allowed: budget.allowed,
            warning: budget.warning,
            spent_usd: budget.spent,
            limit_usd: budget.limit,
            remaining_usd: budget.remaining,
        },
        today: TodayView {
            calls: today.calls,
            tweets_read: today.tweets_read,
            total_cost_usd: today.total_cost,
        },
        reliability,
    };

    if args.json {
        format::print_json_pretty_filtered(&report)?;
        return Ok(());
    }

    println!(
        "{} xint Health ({})",
        icon(report.overall_status),
        report.checked_at
    );
    println!(
        "- Auth status: {} {}",
        icon(report.auth.overall_status),
        status_label(report.auth.overall_status)
    );
    println!(
        "- Budget: {} (${:.2} / ${:.2})",
        if report.budget.allowed {
            "within limit"
        } else {
            "exceeded"
        },
        report.budget.spent_usd,
        report.budget.limit_usd
    );
    println!(
        "- Reliability ({}d): {:.1}% success across {} calls",
        report.reliability.window_days,
        report.reliability.success_rate * 100.0,
        report.reliability.total_calls
    );

    let mut top_commands: Vec<_> = report.reliability.by_command.iter().collect();
    top_commands.sort_by(|a, b| b.1.calls.cmp(&a.1.calls));

    if !top_commands.is_empty() {
        println!("- Top commands:");
        for (command, stats) in top_commands.into_iter().take(5) {
            println!(
                "  {}: {:.1}% ok, p95 {:.0}ms, fallback {:.1}%",
                command,
                stats.success_rate * 100.0,
                stats.p95_latency_ms,
                stats.fallback_rate * 100.0
            );
        }
    }

    Ok(())
}
